### ISSUE TITLE
[Epic][Pdr bot] New smart contract, so user needs just one tx for 2-sided predictions across >1 feeds

### ISSUE THREAD

**Issue by trizin** [oceanprotocol/pdr-backend#701]
Build a smart contract to be able to submit two sided predictions to multiple feeds in a single tx.

### TODO:
- [x] Write the smart contract
- [x] Write tests for the smart contract
- [x] Add PredictionManager module into pdr-backend
- [x] Add a command into pdr-backend to deploy the manager contract.
- [x] Add tests for PredictionManager module
- [x] Update Predictoor PP to support "train_on", "feeds" structure for feeds.
- [x] Implement PredictionManager into predictoor agent

### Flow:
- if doesn't exist, pdr-backend deploys the prediction manager contract using `PRIVATE_KEY` and saves the address of the deployed contract in a file.
- pdr-backend interacts with the smart contract to submit predictions and claim payouts.

### Appendix: Background, motivation, details

(From slack conversation between Berkay & Trent 20240229)

_Berkay:_

Some predictoors use smart contracts, submitting 20 predictions in a single tx.
- Pros: no risk of losing stake due to failed tx, single tx to submit all, single predictoor private key.
- Cons: predictoors gotta deploy & manage a smart contract 

_Trent:_ That's cool! Where did you hear about this?

_Berkay:_ I inspected the top addresses in autobotocean

_Trent:_ We could even consider making this easy for people to do, ie deploy a contract that does predictions across 20 feeds, two-sided

_Berkay:_

Probably yes! we can abstract the complexity. Users would have to not lose their private key and the deployed contract address, and input them to pdr-backend. However the predictoors I've seen were 50-50s, is it feasible to compute the prediction for 10 feeds in a single pdr-backend process?

**Contract interface would look like:**
- `submit([]contracts, []stakesup, []stakesdown, epoch)`**: submit predictions for given epoch to contracts**
- `payout([]contracts, []epochs)`**: claim payout and send funds back to owner**

_Trent:_

Yes, exactly!

Perhaps make a github issue and we can then decide when to do.

If you want to do it sooner rather than later, go for it 

Related : if that existed then we could easily modify predictoor_agent.py to work across >1 feeds.

_Berkay:_

[Issue created!]

It's a huge friction remover and ensures stakes submitted on each side. 

But there's a challenge, calculating prediction for multiple pairs before the epoch closes. We could leverage threading maybe to run the calculations in parallel. Do you think this could be a problem? 

Should we ensure this is working before proceeding with 701?

_Trent_:

Re speed: It comes down to how quickly a model can  be built and predicted from. Right now they're very simple, so very fast. We might even be able to get away with a single thread.

A single thread would be the obvious place to start. Then we can optimize later, if we need.

_Berkay:_

Sounds good!

Current version of predictoor is not bad, I've been using it with our predictoors, caused some stake to be lost in the beginning but didn't have any problems since I set the cut off seconds to 40. But yes, setting up k8s is still a lot of friction.

I can start working on the smart contract solution if you approve, right after I'm done with the current tasks. or I can take another task on the ml side

_Trent:_

Sure do go ahead with the smart contract solution.

BTW I think should keep the K8S functionality too. Anyone doing fancier modeling will need it.

**Comment by trentmc** [comment]
cc @jfdelgad

**Comment by trentmc** [comment]
@trizin fyi Jaime built something like this back in December. He's now moving to work on it again. You guys should sync

**Comment by trentmc** [comment]
Congrats!! Great work
